src/main/java/sorald/processor/StringLiteralInsideEqualsProcessor.java

```diff
@@ -0,0 +1,77 @@
+import spoon.reflect.code.CtExpression;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtMethod;
+import spoon.reflect.declaration.CtType;
+import spoon.reflect.reference.CtExecutableReference;
+
+@ProcessorAnnotation(
+        key = 1132,
+        description =
+                "Strings literals should be placed on the left side when checking for equality")
+public class StringLiteralInsideEqualsProcessor extends SoraldAbstractProcessor<CtInvocation<?>> {
+
+    @Override
+    protected void repairInternal(CtInvocation<?> element) {
+        if (element.getExecutable().getSignature().equals("equals(java.lang.Object)")
+                || element.getExecutable()
+                        .getSignature()
+                        .equals("equalsIgnoreCase(java.lang.String)")) {
+            CtType stringClass = getFactory().Class().get(String.class);
+            CtMethod ctMethodToBeCalled = null;
+            if (element.getExecutable().getSignature().equals("equals(java.lang.Object)")) {
+                ctMethodToBeCalled = (CtMethod) stringClass.getMethodsByName("equals").get(0);
+            } else {
+                ctMethodToBeCalled =
+                        (CtMethod) stringClass.getMethodsByName("equalsIgnoreCase").get(0);
+            }
+            CtExecutableReference ctExecutableReferenceToMethodToBeCalled =
+                    getFactory().Executable().createReference(ctMethodToBeCalled);
+
+            CtInvocation newInvocation =
+                    getFactory()
+                            .Code()
+                            .createInvocation(
+                                    element.getArguments().get(0),
+                                    ctExecutableReferenceToMethodToBeCalled,
+                                    element.getTarget());
+
+            element.replace(newInvocation);
+
+            // the following is to handle the case in which there is a null check on the variable
+            // used as target
+            CtExpression<?> variable = (CtExpression<?>) newInvocation.getArguments().get(0);
+            Boolean nullCheck = false;
+            CtElement parent = element.getParent();
+            CtBinaryOperator parentBinaryOperator = null;
+            while (parent instanceof CtBinaryOperator && !nullCheck) {
+                parentBinaryOperator = (CtBinaryOperator) parent;
+                CtElement parentLeftHandOperand = parentBinaryOperator.getLeftHandOperand();
+                if (parentLeftHandOperand instanceof CtBinaryOperator
+                        && isNullCheckOnTheVariable(
+                                (CtBinaryOperator) parentLeftHandOperand, variable)) {
+                    nullCheck = true;
+                } else {
+                    parent = parent.getParent();
+                }
+            }
+            if (nullCheck) {
+                parentBinaryOperator.replace(newInvocation);
+            }
```

Comment:
The loop is necessary for cases like `(myString != null && (myString != "AAA" && myString.equalsIgnoreCase("foo")))`. I will add a test case for it.

Reply:
My idea was right but there was a bug in the code. I'm pushing both the test and the fix.

Reply:
Okay, I get where you're coming from now. But I don't think this is a good idea, it somewhat goes out of scope for the repair. Let me provide an example that will go horribly wrong:

```java
(myString != null && (myString.startsWith("f") && myString.equalsIgnoreCase("foo")))
```

With the current implementation, this repairs to:

```java
(myString.startsWith("f") && "foo".equalsIgnoreCase(myString))
```

Obviously, this is now open to an NPE on the first call. Now you may be thinking "well I'll just check if theres a method call on the `myString` variable". Well, what about this then:

```java
(myString != null && (someMethodYouCantResolve(myString) && myString.equalsIgnoreCase("foo")))
```

Again, removing the null check opens up for an NPE inside that method (which may or may not call a method on `myString`, we can't tell. Another example would be when a method call on `myString` occurs at the same level as the flagged invocation, or deeper in the binary operator tree, because the current code only checks parents. So this would also be a problem:

```java
(myString != null && (myString.equalsIgnoreCase("foo") || myString.isEmpty())
```

There are countless variations to this. To be safe, we'd really need to remove the null check, and then run Sonar and see if there's now a potential null dereference of `myString`. It's a lot for this single repair.

IMO, the only case where we should ignore a null check is if the entire expression has precisely this form (potentially with `null` and `myString` swapped around in the null check):

```java
myString != null && myString.equals("foo")
```

Reply:
I figured I'll wait with a full review until we've sorted this out :)

Reply:
Ready for review again.